File: addons/tr03112/src/main/java/org/openecard/sal/protocol/eac/gui/CVCStep.java
/****************************************************************************
 * Copyright (C) 2012-2014 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 ***************************************************************************/

package org.openecard.sal.protocol.eac.gui;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import org.openecard.common.I18n;
import org.openecard.gui.definition.Step;
import org.openecard.gui.definition.Text;
import org.openecard.gui.definition.ToggleText;
import org.openecard.sal.protocol.eac.EACData;


/**
 * CVC GUI step for EAC.
 *
 * @author Tobias Wich <tobias.wich@ecsec.de>
 * @author Moritz Horsch <horsch@cdc.informatik.tu-darmstadt.de>
 */
public class CVCStep extends Step {

    // step id
    public static final String STEP_ID = "PROTOCOL_EAC_GUI_STEP_CVC";
    // GUI translation constants
    private static final String TITLE = "step_cvc_title";
    private static final String DESCRIPTION = "step_cvc_description";
    private static final String SUBJECT_NAME = "cvc_subject_name";
    private static final String SUBJECT_URL = "cvc_subject_url";
    private static final String TERMS_OF_USAGE = "cvc_termsofusage";
    private static final String VALIDITY = "cvc_validity";
    private static final String VALIDITY_FORMAT = "cvc_validity_format";
    private static final String VALIDITY_FROM = "cvc_validity_from";
    private static final String VALIDITY_TO = "cvc_validity_to";
    private static final String ISSUER_NAME = "cvc_issuer_name";
    private static final String ISSUER_URL = "cvc_issuer_url";
    private static final String TRANSACTION_INFO = "transaction_info";

    private final I18n lang = I18n.getTranslation("eac");
    private final EACData eacData;

    public CVCStep(EACData eacData) {
	super(STEP_ID);
	this.eacData = eacData;
	setTitle(lang.translationForKey(TITLE));
	setDescription(lang.translationForKey(DESCRIPTION));

	// create step elements
	addElements();
    }

    private void addElements() {
	Text description = new Text();
	description.setText(lang.translationForKey(DESCRIPTION));
	getInputInfoUnits().add(description);

	// SubjectName
	ToggleText subjectName = new ToggleText();
	subjectName.setTitle(lang.translationForKey(SUBJECT_NAME));
	subjectName.setText(eacData.certificateDescription.getSubjectName());
	getInputInfoUnits().add(subjectName);

	// SubjectURL
	ToggleText subjectURL = new ToggleText();
	subjectURL.setTitle(lang.translationForKey(SUBJECT_URL));
	subjectURL.setText(eacData.certificateDescription.getSubjectURL());
	getInputInfoUnits().add(subjectURL);

	// TermsofUsage
	ToggleText termsOfUsage = new ToggleText();
	termsOfUsage.setTitle(lang.translationForKey(TERMS_OF_USAGE));
	termsOfUsage.setText(eacData.certificateDescription.getTermsOfUsage().toString());
	termsOfUsage.setCollapsed(true);
	getInputInfoUnits().add(termsOfUsage);

	// Validity
	DateFormat dateFormat;
	try {
	    dateFormat = new SimpleDateFormat(lang.translationForKey(VALIDITY_FORMAT));
	} catch (IllegalArgumentException e) {
	    dateFormat = new SimpleDateFormat();
	}
	StringBuilder sb = new StringBuilder(150);
	sb.append(lang.translationForKey(VALIDITY_FROM));
	sb.append(" ");
	sb.append(dateFormat.format(eacData.certificate.getEffectiveDate().getTime()));
	sb.append(" ");
	sb.append(lang.translationForKey(VALIDITY_TO));
	sb.append(" ");
	sb.append(dateFormat.format(eacData.certificate.getExpirationDate().getTime()));

	ToggleText validity = new ToggleText();
	validity.setTitle(lang.translationForKey(VALIDITY));
	validity.setText(sb.toString());
	validity.setCollapsed(true);
	getInputInfoUnits().add(validity);

	// IssuerName
	ToggleText issuerName = new ToggleText();
	issuerName.setTitle(lang.translationForKey(ISSUER_NAME));
	issuerName.setText(eacData.certificateDescription.getIssuerName());
	issuerName.setCollapsed(true);
	getInputInfoUnits().add(issuerName);

	// IssuerURL
	ToggleText issuerURL = new ToggleText();
	issuerURL.setTitle(lang.translationForKey(ISSUER_URL));
	issuerURL.setText(eacData.certificateDescription.getIssuerURL());
	issuerURL.setCollapsed(true);
	getInputInfoUnits().add(issuerURL);

	// TransactionInfo
	String transactionInfo = eacData.transactionInfo;
	if (transactionInfo != null) {
	    ToggleText transactionInfoField = new ToggleText();
	    transactionInfoField.setTitle(lang.translationForKey(TRANSACTION_INFO));
	    transactionInfoField.setText(transactionInfo);
	    transactionInfoField.setCollapsed(true);
	    getInputInfoUnits().add(transactionInfoField);
	}
    }

}

File: clients/richclient/src/main/kotlin/org/openecard/richclient/tr03124/ui/CVCStep.kt
/*
 * Copyright (C) 2025 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 */

package org.openecard.richclient.tr03124.ui

import kotlinx.datetime.LocalDate
import kotlinx.datetime.format
import kotlinx.datetime.format.FormatStringsInDatetimeFormats
import kotlinx.datetime.format.byUnicodePattern
import org.openecard.addons.tr03124.eac.EacUiData
import org.openecard.i18n.I18N
import org.openecard.richclient.processui.definition.Document
import org.openecard.richclient.processui.definition.Step
import org.openecard.richclient.processui.definition.Text
import org.openecard.richclient.processui.definition.ToggleText
import org.openecard.sc.pace.cvc.CvcDate.Companion.toLocalDate
import org.openecard.sc.pace.cvc.TermsOfUse

/**
 * CVC GUI step for EAC.
 *
 * @author Tobias Wich
 * @author Moritz Horsch
 * @author Hans-Martin Haase
 */
class CVCStep(
	private val eacData: EacUiData,
) : Step(id = STEP_ID, title = I18N.strings.eac_step_cvc_title.localized()) {
	init {
		description = I18N.strings.eac_step_cvc_description.localized()

		// create step elements
		addElements()
	}

	@OptIn(FormatStringsInDatetimeFormats::class)
	private fun addElements() {
		val description = Text()
		description.text = I18N.strings.eac_step_cvc_description.localized()
		inputInfoUnits.add(description)

		// SubjectName
		val subjectName = ToggleText()
		subjectName.id = "SubjectName"
		subjectName.title = I18N.strings.eac_cvc_subject_name.localized()
		subjectName.text = eacData.certificateDescription.subjectName
		inputInfoUnits.add(subjectName)

		// SubjectURL
		val subjectURL = ToggleText()
		subjectURL.id = "SubjectURL"
		subjectURL.title = I18N.strings.eac_cvc_subject_url.localized()
		subjectURL.text = eacData.certificateDescription.subjectUrl ?: ""
		inputInfoUnits.add(subjectURL)

		// TermsOfUsage
		val termsOfUsage = ToggleText()
		termsOfUsage.id = "TermsOfUsage"
		termsOfUsage.title = I18N.strings.eac_cvc_termsofusage.localized()
		termsOfUsage.document =
			when (val tou = eacData.certificateDescription.termsOfUse) {
				is TermsOfUse.Html -> Document("text/html", tou.html.encodeToByteArray())
				is TermsOfUse.Pdf -> Document("application/pdf", tou.pdf)
				is TermsOfUse.PlainText -> Document("text/plain", tou.text.encodeToByteArray())
			}
		termsOfUsage.isCollapsed = true
		inputInfoUnits.add(termsOfUsage)

		// Validity
		val dateFormat = LocalDate.Format { this.byUnicodePattern(I18N.strings.eac_cvc_validity_format.localized()) }
		val sb = StringBuilder(150)
		sb.append(I18N.strings.eac_cvc_validity_from.localized())
		sb.append(" ")
		sb.append(
			eacData.terminalCert.validFrom
				.toLocalDate()
				.format(dateFormat),
		)
		sb.append(" ")
		sb.append(I18N.strings.eac_cvc_validity_to.localized())
		sb.append(" ")
		sb.append(
			eacData.terminalCert.validUntil
				.toLocalDate()
				.format(dateFormat),
		)

		val validity = ToggleText()
		validity.id = "Validity"
		validity.title = I18N.strings.eac_cvc_validity.localized()
		validity.text = sb.toString()
		validity.isCollapsed = true
		inputInfoUnits.add(validity)

		// IssuerName
		val issuerName = ToggleText()
		issuerName.id = "IssuerName"
		issuerName.title = I18N.strings.eac_cvc_issuer_name.localized()
		issuerName.text = eacData.certificateDescription.issuerName
		issuerName.isCollapsed = true
		inputInfoUnits.add(issuerName)

		// IssuerURL
		val issuerURL = ToggleText()
		issuerURL.id = "IssuerURL"
		issuerURL.title = I18N.strings.eac_cvc_issuer_url.localized()
		// issuer url is optional so perform a null check
		issuerURL.text = eacData.certificateDescription.issuerUrl ?: ""
		issuerURL.isCollapsed = true
		inputInfoUnits.add(issuerURL)
	}

	companion object {
		const val STEP_ID: String = "PROTOCOL_EAC_GUI_STEP_CVC"
	}
}
